lib/patientFinder/utils.rb

```diff
@@ -0,0 +1,165 @@
+        def self.get_candidates_for_test(master_records,test_record)
+            
+            candidates=Array.new
+            # candidates
+            master_records.each { |master_record|
+            
+                #simple conditions that will reduce the record pool
+                if(master_record.gender==test_record.gender && master_record.ethnicity[:code]==test_record.ethnicity[:code] && master_record.race[:code] == test_record.race[:code])
+                    
+                    #go inside the sections and collect codes and find if the codes from the test are ALL contained in the the same section of the master 
+                    # sections = [:allergies, :care_goals, :conditions, :encounters, :immunizations, :medical_equipment, :medications, :procedures, :insurance_providers ]
+                
+                    # binding.pry
+                    #TODO - couldn't get this to work on an elegant way - further work needed in order to make it nice
+                    # get_non_empty_sections_for_record(sections, test_record).each{ |sec|
+                    # puts "#{sec}"
+                    # }
+                    # sections_with_data=get_non_empty_sections_for_record(sections, test_record)
+                    
+                    # master_is_match = sections_with_data.all? {|section| Utils::codes_match?(test_record[section],master_record[section])}
+                    
+                    # if(master_is_match)
+                    #     candidates<< master_record
+                    # end
+                    
+                    allergiesMatch = codes_match?(test_record.allergies,master_record.allergies)
+                    care_goalsMatch = codes_match?(test_record.care_goals,master_record.care_goals)
+                    conditionsMatch = codes_match?(test_record.conditions,master_record.conditions)
+                    encountersMatch = codes_match?(test_record.encounters,master_record.encounters)
+                    immunizationsMatch = codes_match?(test_record.immunizations,master_record.immunizations)
+                    medical_equipmentMatch = codes_match?(test_record.medical_equipment,master_record.medical_equipment)
+                    medicationsMatch = codes_match?(test_record.medications,master_record.medications)
+                    proceduresMatch = codes_match?(test_record.procedures,master_record.procedures)
+                    insurance_providersMatch = codes_match?(test_record.insurance_providers,master_record.insurance_providers)
+    
+                    #create candidates based on matching codes - if all the above match then consider this master as candidate
+                    if(allergiesMatch && care_goalsMatch && conditionsMatch && encountersMatch &&  immunizationsMatch && medical_equipmentMatch &&  medicationsMatch && proceduresMatch &&  insurance_providersMatch)
```

Comment:
If you were to continue doing it this way, a slightly nicer way of doing this would be:

    matches = [allergiesMatch, care_goalsMatch, conditionsMatch, encountersMatch, immunizationsMatch, medical_equipmentMatch, medicationsMatch, proceduresMatch, insurance_providersMatch]
    if (matches.all?)
        candidates << master_record
    end

It could then be made even nicer by replacing each assignment with a shovel onto the array, i.e.

    matches = []
    matches << codes_match?(test_record.allergies, master_record.allergies)
    matches << codes_match?(test_record.care_goals, master_record.care_goals)
    ...
    matches << codes_match?(test_record.insurance_providers, master_record.insurance_providers)
    
    if (matches.all?)
        candidates << master_record
    end

I still prefer your original way with one of the fixes mentioned above, though.